proposals/lip-0053.md

```diff
@@ -133,28 +144,247 @@ The following constants are used throughout the document:
-| `CHAIN_ACTIVE`                                           | uint32 | 1                       |  
-| `CHAIN_TERMINATED`                                       | uint32 | 2                       |
+| `CHAIN_STATUS_REGISTERED`                                | uint32 | 0                       |
+| `CHAIN_STATUS_ACTIVE`                                    | uint32 | 1                       |  
+| `CHAIN_STATUS_TERMINATED`                                | uint32 | 2                       |
 | **Message Tags**                                         |        |                         |
 | `MESSAGE_TAG_CERTIFICATE `                               | bytes  | "LSK_CE_" ASCII-encoded |
 
 Several of those constants are shared with the other LIPs defining the Interoperability module and all of the needed constants for the Interoperability module are defined in [LIP 0045][lip-0045]. That LIP should be considered correct if a value stated here differs.
 
-#### uint32be Function
+### Type Definition
+
+| Name               | Type    | Validation            | Description                     |
+|--------------------|---------|-----------------------|---------------------------------|
+| `Address` | bytes   | Must be of length `ADDRESS_LENGTH`. | Address of an account.   |
+| `ModuleID` | bytes   | Must be of length `MODULE_ID_LENGTH`. | ID of a module.   |
+| `CrossChainCommandID` | bytes   | Must be of length `CROSS_CHAIN_COMMAND_ID_LENGTH`. | ID of a cross-chain command.   |
+| `ChainID` | bytes   | Must be of length `CHAIN_ID_LENGTH`. | ID of a chain.   |
+| `CCM` | object   | Must follow the `crossChainMessageSchema` schema. | The type of cross-chain messages.   |
+| `MerkleRoot` | bytes   | Must be of length `MERKLE_ROOT_LENGTH`. | The root of a Merkle tree.   |
+| `PublicKeyEd25519` | bytes | Must be of length `ED25519_PUBLIC_KEY_LENGTH`. | Used for Ed25519 public keys. |
+| `PublicKeyBLS` | bytes | Must be of length `BLS_PUBLIC_KEY_LENGTH`. | Used for BLS keys. |
+| `ActiveValidator` | object | Must have `blsKey: PublicKeyBLS` and `bftWeight: uint64` properties. | The BLS key and BFT weight of an active validator. |
+
+### Auxiliary Functions
+
+The following auxiliary functions are used internally by the Interoperability module to verify and execute cross-chain updates.
+
+#### verifyLivenessCondition
+
+```python
+def verifyLivenessCondition(trs: Transaction) -> None:
+    sidechainID = trs.params.sendingChainID
+
+    if chainAccount(sidechainID).status == CHAIN_STATUS_REGISTERED:
+        certificateBytes = trs.params.certificate
+        if len(certificateBytes) == 0:
+            raise Exception("The first mainchain CCU must contain a non-empty certificate.")
+        
+        certificate = decode(certificateSchema, certificateBytes)
+        timestamp = timestamp of the block including CCU
+        if len(trs.params.inboxUpdate) > 0 and timestamp - certificate.timestamp > LIVENESS_LIMIT / 2:
```

Comment:
I am not completely sure that understand why the check `en(trs.params.inboxUpdate) > 0` is necessary. At least it is not mentioned in the rationale above.

I assume that we want to allow submitting multiple certificates without inbox update until the certificate is at most 15 days old.

Reply:
The point is that if the inboxUpdate is empty, it will not contain the registration CCM that will activate the chain status. So the following check is only needed if the ccu will activate the chain status and the certificate timestamp is too close to the current timestamp

Reply:
I don't understand the explanation.

Reply:
if the `inboxUpdate` is empty, it implies that the CCU does not contain any CCM (but only the certificate). In particular, this means that it will not contain the registration CCM that will activate the chain status (as this is supposed to be the very first ccm that gets posted). The check on the liveness is then done to ensure that the first CCU that activates the chain does not bring it immediately too close to the termination threshold (but we still have this 15 days of extra time)

Reply:
Thanks for the explanation. If I understand correctly then, the activation for a blockchain would work as follows:

1) Submit CCUs with certificates to update the validators up to some very recent point in time (maybe the registration and activation are several month apart)
2) Submit CCU with registration CCM for activating the chain.

If this is correct, then the error message
> "First certificate cannot be older than {LIVENESS_LIMIT / 2} seconds.
is a bit misleading.

Reply:
true, I updated it

Reply:
Looks good now.